### ISSUE TITLE
asciidoctor gradle plugin not rendering inline macro extension

### ISSUE THREAD

**Issue by efeig** [asciidoctor/asciidoctor-gradle-plugin#203]
## Expected Behavior

I use the extensions configuration method of the asciidoctor gradle plugin to add an inline macro, that transforms the following markup `issue:PRJ-100[]` into a link that points to the web page for issue PRJ-100 (see build file below). Until gradle version 2.14 the result was as expected: In the generated html file all occurences were replaced by a proper link.

## Current Behavior

In gradle version 3.1 and 3.2.1 the inline macro is not rendered correctly any more. After switching to a new gradle version the link is created correctly only once. In following builds the pattern `issue:PRJ-100[]` is not transformed into a link any more. I realized that the problem only occurs when I include the line

     classpath 'org.asciidoctor:asciidoctorj-pdf:1.5.0-alpha.11'

without generating a pdf file (i.e. using pdf as backend). Leaving the line as it is and using pdf as backend or commenting the line out leads to the expected result.

## Context

In principle it makes no sense to include

     classpath 'org.asciidoctor:asciidoctorj-pdf:1.5.0-alpha.11'

wihtout generating pdf files from AsciiDoc files. However, should this line affect the correct rendering of inline macros in html files?

## Steps to Reproduce (for bugs)

My build file looks like that
```
buildscript {
    repositories {
        jcenter()
    }

    dependencies {
        classpath 'org.asciidoctor:asciidoctor-gradle-plugin:1.5.3'
        classpath 'org.asciidoctor:asciidoctorj-pdf:1.5.0-alpha.11'
    }
}

apply plugin: 'org.asciidoctor.convert'

asciidoctor {
    backends=['html5']
//    backends=['html5', 'pdf']
    extensions {
         inlinemacro(name: "jira") {
            parent, target, attributes ->
              options = [
                "type": ":link",
                "target": "https://my.issue.server/${target}".toString()
            ]
            createInline(parent, "anchor", target, attributes, options).render()
        }
      }    
}
```
Putting a test.adoc file in the default sources dir `src/docs/asciidoc` with following content
```
jira:PRJ-1058[] 
```
and executing the asciidoctor task, should reproduce the problem.

## Your Environment

* Build scan URL: https://scans.gradle.com/s/hb2x6qnjapvfg/infrastructure

**Comment by abelsromero** [comment]
I tested it and I found the problem, but not a nice solution.
**First**, good news, **I don't think it is related with `asciidoctorj-pdf`** since I reproduced with and without it.

Then, **quick & dirty work-arround**.
Version 3.x made the gradle daemon start a new instance by default on a first run, that's when it works, successive executions run on the daemon, those executions fail.
If you stop the daemon always, the extension always works:
```
./gradlew --stop && ./gradlew clean asciidoctor
```
This basically makes gradle work as in v2.x. For mor info -> https://docs.gradle.org/current/userguide/gradle_daemon.html.

**Thew issue**, I'd try to explain even when I don't fully understand it.
The extension are part of the groovy DSL and this is loaded using [AsciidoctorExtensions](https://github.com/asciidoctor/asciidoctorj-groovy-dsl/blob/master/src/main/groovy/org/asciidoctor/groovydsl/AsciidoctorExtensions.groovy) component.
Long story short:
1. the extensions are stored in the static attribute `REGISTERED_EXTENSIONS` (in this case, the extensio is a Closure), 

2. later, during Asciidoctor initialitzation the `registerTo` method is invoked to register the extensions in Asciidoctor's instance.

However, debugging I showed that the instance of REGISTERED_EXTENSIONS invoked at step 1 and 2 is not same, the second time, it always invokes the first one that was created during the first run. That's why the first time works but on successive invocations it doesn't. Step 1 stores the extension in a new instance everytime, but the registerTo always checks the initial, so, expect for the first run, it always finds 0 pending extensions to register.
Why is that?..I don't know, but the plugin replaces the Thread classloader to dynamically load Asciidoctorj jars. However, this is done previously to any instantiation, so to my understanding, everything should be new on every run. At least the Asciidoctor and the Task instance are.

Possible solutions I can think off, but not sure if they work:
· Change how we load the dependencies. I remember reading about it here https://leanpub.com/idiomaticgradle (legally owned). @ysb33r can probably shed some light here.
· Create the Asciidoctor instance passing the classloader. This is an option, but not sure how this will impact when gemPaths option is used. Most likely it will fail :/
· Change how the DSL is registered to avoid static variables... not sure this is even possible since the DSL uses the SPI interface.

**Comment by ysb33r** [comment]
Wil shot here, but this is what is being used to load extensions. `private static ClassLoader cl`. I am not sure what the exact details are as to how the daemon load new projects, but I am suspectintg the `cl` is inly initialised once and that is what you will get on subsequent runs.

**Comment by abelsromero** [comment]
I’ve been looking into this, and I don’t think there’s a solution that fixes the issue AND also  allows selecting the AsciidoctorJ/DSL version inside the Gradle Daemon.

I have confirmed that if no changes are done in the build script, the Gradle Daemon uses the same thread with the same classloader. So apparently creating a new classloader that gets discarded after the task should work. But, as explained in the previous comment, the original instance is retrieved, even when that classloader is not referenced in the current thread classloader.
This could be due to a leak in AsciidoctorJ or jRuby, but forcing a garbage collection does not remove old class definitions from the heap. I always assumed that to be normal, but I also read that Java8 introduced some changes, still, I haven’t been able to find clear documentation on the matter. So, my assumption is that once a class is loaded it remains and no matter the classloader at that moment, that class is used from that moment on.

This has a side-effect, because the same problem applies  to `Asciidoctor` class , this does not impact execution. But it means that, after the first execution, changing Asciidoctorj version in the extension has no effect at all.
Based on that, my *current* proposal is to:
-	Fix the current issue with extensions not loading: just initializing `private static ClassLoader cl` once instead of on every execution (https://github.com/asciidoctor/asciidoctor-gradle-plugin/blob/development/src/main/groovy/org/asciidoctor/gradle/AsciidoctorTask.groovy#L844). A simple null validation is enough. Btw, I tried making `cl` non static but the issue remains.
-	Document in README that changes in AsciidoctorJ or the DSL versions require to restart the Daemon. Optionally we can check that and show a warning message also during excecution.

As a side note, when the build script changes, the Daemon launches a new thread. In both cases a new Task is created.

I will also check in the forums and see what Team says. The fact that they provide this feature https://docs.gradle.org/2.5/release-notes#simpler-default-dependencies makes me think there should be a way to cleanly load dependencies.

**Comment by ysb33r** [comment]
> I will also check in the forums and see what Team says. The fact that they provide this feature https://docs.gradle.org/2.5/release-notes#simpler-default-dependencies makes me think there should be a way to cleanly load dependencies.

This ony provide the dependency. It still has to be loaded.

Anyhow I think it is worthwhile posting to the forum referencing this issue.

**Comment by mojavelinux** [comment]
> Document in README that changes in AsciidoctorJ or the DSL versions require to restart the Daemon.

As a user, that seems like a reasonable restriction to me. (Of course, if you can figure out how to lift that restriction, then go for it).

**Comment by oehme** [comment]
Restarting the daemon should not be considered normal or acceptable. I can only recommend using a profiler to find out what is keeping a strong reference to that classloader.

**Comment by abelsromero** [comment]
> Restarting the daemon should not be considered normal or acceptable.

I am still trying to find the good solution. I was able to write a separated code that replicated the logic of the plugin and worked, so it is definitely possible.

**Comment by oehme** [comment]
I just tried the following snippet: 

```
asciidoctor {
    backends=['html5']
    extensions {
        println delegate.class.classLoader
     }
}
```

It's a different classloader on each invocation, as expected. Can you please provide a reproducible example project?

**Comment by oehme** [comment]
Ah, I see what's happening here now: @efeig added the pdf backend to the `buildscript.classpath`, which means that the asciidoctor configuration is no longer taken into account, because the build script classpath is the parent and thus classes are loaded from there first.

If I understand the plugin correctly, what you should have done instead is 

```
dependencies {
  asciidoctor 'org.asciidoctor:asciidoctorj-pdf:1.5.0-alpha.11'
}
```

@abelsromero you might want to throw an excpetion in your plugin when someone leaks asciidoctor into the plugin's classpath.

**Comment by abelsromero** [comment]
After testing it, I can confirm that was the issue all from the beginning. 
It happens because `Asciidoctorj-pdf` has a `runtime` dependency with `Asciidoctorj`, so Gradle loads the library before initializing the plugin.
@efeig the proper way to load external extensions libraries is using the `asciidoctor` configuration like in here https://github.com/asciidoctor/asciidoctor-gradle-plugin#as-external-library.

Honestly, I feel kind of dumb for not seeing it before, the documentation is clear. However I think this can be improved to avoid miss configurations, first we add a warning section here, then as suggested by @oehme we can check for the class and fail if we detect an instance of Asciidoctor before setting the classloader. wdyt?

Alos, I just noticed the gradle [examples](https://github.com/asciidoctor/asciidoctor-gradle-examples/blob/master/asciidoc-to-pdf-example/build.gradle) use the failing configuration, shouldn't they be fixed too?

**Comment by oehme** [comment]
You should consider having two configurations, one for the asciidoctor core (let's call it `asciidoctorCore`) which will have default dependencies and one for asciidoctor extensions (let's call it just `asciidoctor`) which the user will often customize. The reason is that by adding an extension the user currently overwrites the default dependencies and might end up with an incomplete core. With the suggested split this would no longer happen.

**Comment by ysb33r** [comment]
> You should consider having two configurations, one for the asciidoctor core (let's call it asciidoctorCore) which will have default dependencies and one for asciidoctor extensions (let's call it just asciidoctor) which the user will often customize. The reason is that by adding an extension the user currently overwrites the default dependencies and might end up with an incomplete core. With the suggested split this would no longer happen.

@oehme, have jou seen #201  ?

**Comment by oehme** [comment]
No I didn't see that, thanks for the pointer. But let's not bind this simple improvement to a giant refactoring.

**Comment by oehme** [comment]
I see, I could have sworn I saw `defaultDependencies` being used. Maybe there have been changes since I last looked. Seems like it is (now) exactly as I proposed.

**Comment by efeig** [comment]
Thank you for pointing out the problem to me: using the asciidoctor configuration to load the pdf extension instead of `buildscript.classpath` works perfect.

**Comment by aalmiray** [comment]
Is there anything left to be done for this issue, code-wise?
It looks to me this is a documentation issue.

**Comment by ysb33r** [comment]
Closing this, but still tracking it via #222 as there might be a related issue.
